push-common-updates pipeline failing with `BUILD_BUILDNUMBER` error

```
##[error]TF209010: The build number format $env:BUILD_BUILDNUMBER (Dry run)\"\n        Write-Host \"##vso[build.addbuildtag]dry-run\"\n        \n        if (\"$(officialBranches)\".Split(',').Contains(\"$(Build.SourceBranch)\"))\n        {\n          Write-Host \"##vso[task.logissue type=error]Cannot run a CG dry-run build from an official branch ($(officialBranches)).\"\n          Write-Host \"##vso[task.logissue type=error]Run the pipeline again from a different branch to avoid registering scan results.\"\n          exit 1\n        }\n      displayName: Update Build Number\n    - task: ComponentGovernanceComponentDetection@0\n      displayName: Component Detection (manually injected)\n      inputs:\n        # Running CG with `whatIf: true` or `scanType: LogOnly` outputs a list of detected components, but doesn't show\n        # which components would trigger an alert on a production build.\n        # As long as the build isn't ran on a tracked branch (typically main or nightly), it's OK to submit components\n        # to CG for analysis. Only the results for tracked branches matter.\n        scanType: Register\n        whatIf: false\n        alertWarningLevel: Low\n        failOnAlert: false\n        ignoreDirectories: $(Build.SourcesDirectory)/versions\n        showAlertLink: true\n        timeoutInMinutes: 10\n" contains invalid character(s), is too long, or ends with '.'. The maximum length of a build number is 255 characters. Characters which are not allowed include '"', '/', ':', '<', '>', '\', '|', '?', '@', and '*'.
```

[Example build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2512285&view=results) (internal link)

This is a regression from #1371

--- a/eng/src/file-pusher/AzDoSafeTraceListenerWrapper.cs
+++ b/eng/src/file-pusher/AzDoSafeTraceListenerWrapper.cs
@@ -1,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace FilePusher;
+
+public class AzDoSafeTraceListenerWrapper(TraceListener innerTraceListener) : TraceListener
+{
+    private readonly TraceListener _innerTraceListener = innerTraceListener;
+
+    public override void Write(string? message)
+    {
+        _innerTraceListener.Write(EscapeVsoDirectives(message));
+    }
+
+    public override void WriteLine(string? message)
+    {
+        _innerTraceListener.WriteLine(EscapeVsoDirectives(message));
+    }
+
+    private static string? EscapeVsoDirectives(string? message)
+    {
+        return message?.Replace("##vso", "#VSO_DIRECTIVE");
+    }
+}

--- a/eng/src/file-pusher/AzDoSafeTraceListenerWrapper.cs
+++ b/eng/src/file-pusher/AzDoSafeTraceListenerWrapper.cs
@@ -20,6 +20,13 @@ public class AzDoSafeTraceListenerWrapper(TraceListener innerTraceListener) : Tr
         _innerTraceListener.WriteLine(EscapeVsoDirectives(message));
     }
 
+    /// <summary>
+    /// This method "escapes" Azure DevOps Pipeline tasks/variable assignments in strings so that they are safe to
+    /// output to the console in pipeline.
+    /// This prevents issues like https://github.com/dotnet/docker-tools/issues/1388, where pushing files containing
+    /// these AzDO variable assignments results in pipeline failure.
+    /// Azure DevOps documentation: https://learn.microsoft.com/en-us/azure/devops/pipelines/process/set-variables-scripts
+    /// </summary>
     private static string? EscapeVsoDirectives(string? message)
     {
         return message?.Replace("##vso", "#VSO_DIRECTIVE");

--- a/eng/src/file-pusher/FilePusher.cs
+++ b/eng/src/file-pusher/FilePusher.cs
@@ -33,7 +33,9 @@ namespace FilePusher
             try
             {
                 // Hookup a TraceListener to capture details from Microsoft.DotNet.VersionTools
-                Trace.Listeners.Add(new TextWriterTraceListener(Console.Out));
+                TextWriterTraceListener textWriterTraceListener = new(Console.Out);
+                AzDoSafeTraceListenerWrapper safeTraceListener = new(textWriterTraceListener);
+                Trace.Listeners.Add(safeTraceListener);
 
                 string configJson = File.ReadAllText(options.ConfigPath);
                 Config config = JsonConvert.DeserializeObject<Config>(configJson)

--- a/eng/src/yaml-updater/YamlUpdater.cs
+++ b/eng/src/yaml-updater/YamlUpdater.cs
@@ -9,6 +9,7 @@ using System.Diagnostics;
 using System.IO;
 using System.Text;
 using System.Threading.Tasks;
+using FilePusher;
 using Newtonsoft.Json;
 using YamlDotNet.Core;
 using YamlDotNet.Core.Events;
@@ -35,7 +36,9 @@ namespace YamlUpdater
         private static async Task ExecuteAsync(Options options)
         {
             // Hookup a TraceListener to capture details from Microsoft.DotNet.VersionTools
-            Trace.Listeners.Add(new TextWriterTraceListener(Console.Out));
+            TextWriterTraceListener textWriterTraceListener = new(Console.Out);
+            AzDoSafeTraceListenerWrapper safeTraceListener = new(textWriterTraceListener);
+            Trace.Listeners.Add(safeTraceListener);
 
             string configJson = File.ReadAllText(options.ConfigPath);
             FilePusher.Models.Config config = JsonConvert.DeserializeObject<FilePusher.Models.Config>(configJson);

